[inductor] Constant and index_expr propagation pass

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101602
* #100895
* __->__ #101077

This pass does a limited form of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @anijain2305

peterbell10 committed:
[inductor] WIP: index-propagation pass

This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: index-propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: index-propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] WIP: index-propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire anijain2305

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire anijain2305

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -2509,37 +2513,39 @@ def compute_scale(in_size, out_size, align_corners, scale=None):
+            half = ops.constant(0.5, torch.float32)
             return ops.sub(
-                ops.mul(scale, ops.add(dst_index_ie, 0.5)), 0.5
+                ops.mul(scale, ops.add(dst_index_ie, half)), half
```
There lowering changes are just fixing cases where constants aren't being declared with `ops.constant`.

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -4191,7 +4191,7 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
             def index_expr(self, index, dtype):
                 if isinstance(index, (int, sympy.Integer)):
-                    return ops.constant(int(index), dtype)
+                    return self._inner.constant(int(index), dtype)
```
`self._inner` can return a different type from the global ops handler, so we need to be more consistent here.

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -4232,15 +4232,23 @@ def set_indirect(new_var):
+            )
         ):
-            tracer.create_proxy("output", "output", (fn(*args),), {})
+            ops.output(fn(*args))
```
This indirection is just a cute way to get `IndexPropagation` to unwrap the return value.

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -6288,7 +6286,6 @@ def fn():
 
                 # this can be optimized away, value too large
                 self.assertTrue("to(tl.int64)" not in code)
-                self.assertTrue("to(tl.int32)" in code)
```
This is so nice. Triton code is going to be so much easier to read now. Forget having to track down variables that then end up being just a `tmp0 = 1`

lezcano replied:
The issue here (and in the test above) is that they test the value range analysis. Now, with the constant folding, all these are folded. We need to change these examples so that the make sense.

peterbell10 replied:
I've added a config option so we can disable constant propagation for these tests.

lezcano replied:
I wonder whether there's anything left for ValueRanges with static shapes after constant propagation... I think there will be examples where it is still beneficial, but I couldn't name one off the top of my head. Perhaps things with `pow` and weirder ops.

lezcano commented on torch/_inductor/index_propagation.py:
```diff
@@ -0,0 +1,167 @@
+        return TypedExpr(-x.expr, x.dtype)
+
+
+@dataclass
+class IndexPropVar:
+    rep: Any  # A wrapped IR value, if present
+    expr: Optional[TypedExpr] = None
```
Could this be written as
```suggestion
@dataclass
class IndexPropVar:
    rep: Union[TypedExpr, Any] = None  # Any == wrapped IR value
```
?

peterbell10 replied:
In the unlikely event that the wrapped handler returns `TypedExpr`, this would behave unexpectedly.

lezcano replied:
Then let's assert that it doesn't do that and change it to the more compact representation? I believe it's makes easier to understand what's really going on (we can get rid of the `id_iter` and the `index_tmp` name that's never used).

lezcano reviewed (approved):
Really neat implementation! I left a few rather minor comments, but this LGTM really.

As a side note, I thought we had to be careful with overflows when doing constant folding, but then I realised that triton has a limitation with these things. See https://github.com/pytorch/pytorch/issues/101684

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire anijain2305

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/index_propagation.py:
```diff
@@ -0,0 +1,167 @@
+"""
+This file implements the IndexPropagation ops handler, which wraps an
+underlying handler to add a limited for of constant propagation, as well as
```
Added an example and discussion of the expression printer limitations.

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -2509,37 +2513,39 @@ def compute_scale(in_size, out_size, align_corners, scale=None):
         # ((A * x - 5 * A) * x + 8 * A) * x - 4*A
         return ops.sub(
-            ops.mul(ops.add(ops.mul(ops.sub(ops.mul(A, x), 5 * A), x), 8 * A), x), 4 * A
+            ops.mul(ops.add(ops.mul(ops.sub(ops.mul(_A, x), _5A), x), _8A), x), _4A
```
I think having mypy type-check this would be the best solution.

peterbell10 committed:
lint on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire anijain2305

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire anijain2305

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire anijain2305

[ghstack-poisoned]

ngimel reviewed (approved):
This is great!

peterbell10 committed:
Update on "[inductor] Constant and index_expr propagation pass"


This pass does a limited for of constant propagation, as well as propagation of
sympy indexing expressions. For example, say you have the function:
```python
def flip(x):
    i = torch.arange(x.size(0) - 1, -1, -1, device=x.device)
    return x[i]
```

On current main this results in indirect indexing:
```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = 3 - z0
    index1 = 3*indirect0 + z1
    index2 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        set_indirect0 = self.set_indirect0(index_expr)
        get_index_1 = self.get_index('index1')
        load = ops.load('arg0_1', get_index_1)
        get_index_2 = self.get_index('index2')
        store = ops.store('buf0', get_index_2, load, None)
        return store
```

With this PR the indexing is propagated through the computation and into direct
indexing:

```python
class buf0_loop_body:
    var_ranges = {z0: 4, z1: 3}
    index0 = -3*z0 + z1 + 9
    index1 = 3*z0 + z1
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('arg0_1', get_index)
        get_index_1 = self.get_index('index1')
        store = ops.store('buf0', get_index_1, load, None)
        return store
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire anijain2305

[ghstack-poisoned]
